### PR TITLE
unified cache 

### PR DESCRIPTION
**Pull request by SomkaPe**

### CODE REVIEW

File: adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs [hunk 1]
```diff
@@ -315,7 +316,10 @@ protected virtual Task PreTokenRequest()
 
         protected virtual async Task<AdalResultWrapper> SendTokenRequestAsync()
         {
-            var requestParameters = new DictionaryRequestParameters(this.Resource, this.ClientKey);
+            var requestParameters = new DictionaryRequestParameters(this.Resource, this.ClientKey)
+            {
+                { OAuth2Parameter.ClientInfo, "1" }
```
Comment (henrik-me) [inline adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs]: >{ OAuth2Parameter.ClientInfo, "1" } [](start = 16, length = 35)

Why is this not added in AddAdditionalRequestParameters?

File: adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs [hunk 2]
```diff
@@ -438,7 +442,7 @@ private void LogReturnedToken(AdalResult result)
 
         protected void ValidateAuthorityType()
         {
-            if (!SupportADFS && Authenticator.AuthorityType == AuthorityType.ADFS)
+            if (!SupportADFS && Authenticator.AuthorityType == Instance.AuthorityType.ADFS)
```
Comment (henrik-me) [inline adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs]: >Instance [](start = 63, length = 8)

?

File: msal/src/Microsoft.Identity.Client/IUser.cs [hunk 3]
```diff
@@ -47,7 +47,7 @@ public interface IUser
         /// <summary>
         /// Gets identity provider if returned by the service. If not, the value is null.
         /// </summary>
-        string IdentityProvider { get; }
+        //string IdentityProvider { get; }
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/IUser.cs]: >//string IdentityProvider { get; } [](start = 8, length = 34)

remove?

File: msal/src/Microsoft.Identity.Client/User.cs [hunk 4]
```diff
@@ -43,10 +43,10 @@ internal User(User other)
             DisplayableId = other.DisplayableId;
             Identifier = other.Identifier;
             Name = other.Name;
-            IdentityProvider = other.IdentityProvider;
+            //IdentityProvider = other.IdentityProvider;
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/User.cs]: >//Iden [](start = 12, length = 6)

why uncomment and not remove?

File: msal/src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs [hunk 5]
```diff
@@ -51,7 +51,7 @@ public SilentRequest(AuthenticationRequestParameters authenticationRequestParame
         protected override void SetAdditionalRequestParameters(OAuth2Client client)
         {
             client.AddBodyParameter(OAuth2Parameter.GrantType, OAuth2GrantType.RefreshToken);
-            client.AddBodyParameter(OAuth2Parameter.RefreshToken, _msalRefreshTokenItem.RefreshToken);
+            client.AddBodyParameter(OAuth2Parameter.RefreshToken, _msalRefreshTokenItem.Secret);
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs]: >Secret [](start = 88, length = 6)

why this name?

File: msal/tests/dev apps/XForms/XForms/RefreshTokenCacheItemDetails.xaml.cs [hunk 6]
```diff
@@ -39,21 +39,22 @@ namespace XForms
     [XamlCompilation(XamlCompilationOptions.Compile)]
     public partial class RefreshTokenCacheItemDetails : ContentPage
     {
-        internal RefreshTokenCacheItemDetails(MsalRefreshTokenCacheItem msalRefreshTokenCacheItem)
+        internal RefreshTokenCacheItemDetails(MsalRefreshTokenCacheItem msalRefreshTokenCacheItem, 
+            MsalAccountCacheItem MsalAccountCacheItem)
```
Comment (henrik-me) [inline msal/tests/dev apps/XForms/XForms/RefreshTokenCacheItemDetails.xaml.cs]: >M [](start = 33, length = 1)

lower case ? it seems like a static this way or a property.

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs [hunk 7]
```diff
@@ -550,35 +551,58 @@ public void GetUsersTest()
...
                 ExpiresOnUnixTimestamp =
                     CoreHelpers.DateTimeToUnixTimestamp((DateTime.UtcNow + TimeSpan.FromSeconds(3600))),
-                RawIdToken = MockHelpers.CreateIdToken(TestConstants.UniqueId, TestConstants.DisplayableId),
                 RawClientInfo = MockHelpers.CreateClientInfo(),
                 ScopeSet = TestConstants.Scope
             };
-            item.IdToken = IdToken.Parse(item.RawIdToken);
-            item.ClientInfo = ClientInfo.CreateFromJson(item.RawClientInfo);
-            item.AccessToken = item.GetAccessTokenItemKey().ToString();
+            item.CreateDerivedProperties();
```
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs]: >CreateDerivedProperties [](start = 17, length = 23)

why is this needed? Can't the MsalAccessTokenCacheItem detect itself when that is needed (and e.g. ensure it does what ever it needs to do when someone calls a "calculated" property?

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs [hunk 8]
```diff
@@ -550,35 +551,58 @@ public void GetUsersTest()
...
-                RawClientInfo = MockHelpers.CreateClientInfo(TestConstants.Uid + "more", TestConstants.Utid),
-                DisplayableId = TestConstants.DisplayableId,
-                IdentityProvider = TestConstants.IdentityProvider,
-                Name = TestConstants.Name
+                Secret = "someRT",
+                RawClientInfo = MockHelpers.CreateClientInfo("uId1", "uTId1"),
+                TenantId = "uTId1"
             };
-            rtItem.ClientInfo = ClientInfo.CreateFromJson(rtItem.RawClientInfo);
+            rtItem.InitRawClientInfoDerivedProperties();
```
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs]: >InitRawClientInfoDerivedProperties [](start = 19, length = 34)

similar comment to above

File: core/src/Cache/MsalCacheItemBase.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                return null;
+            }
+            return string.Format(CultureInfo.InvariantCulture, "{0}.{1}", ClientInfo.UniqueIdentifier,
+                ClientInfo.UniqueTenantIdentifier);
+        }
+
+        [OnDeserialized]
+        void OnDeserialized(StreamingContext context)
+        {
+            InitClientInfo();
```
Comment (henrik-me) [inline core/src/Cache/MsalCacheItemBase.cs]: >InitClientInfo [](start = 12, length = 14)

InitRawClientInfoDerivedProperties ?

File: core/src/Cache/MsalCacheItemBase.cs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+        internal void InitClientInfo()
+        {
+            if (RawClientInfo != null)
+            {
+                ClientInfo = ClientInfo.CreateFromJson(RawClientInfo);
+            }
+        }
+
+        internal void InitRawClientInfoDerivedProperties()
```
Comment (henrik-me) [inline core/src/Cache/MsalCacheItemBase.cs]: >InitRawClientInfoDerivedProperties [](start = 22, length = 34)

not a fan of someone external having to call this one...

File: core/src/Cache/MsalIdTokenCacheItem.cs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        internal string Authority { get; set; }
+
+        internal IdToken IdToken { get; set; }
+
+        internal void CreateDerivedProperties()
+        {
+            IdToken = IdToken.Parse(Secret);
+
+            InitRawClientInfoDerivedProperties();
+        }
```
Comment (henrik-me) [inline core/src/Cache/MsalIdTokenCacheItem.cs]: should be private....

**Comment by henrik-me** [comment]
> 

I'm surprised that you need this for the core library unit tests. why?

---
Refers to: core/tests/Test.Microsoft.Identity.Core.Unit/Properties/InternalsVisibleTo.cs:37 in 0a4acbb. [](commit_id = 0a4acbb7026204a4698a1b012eb19ec2f45648e6, deletion_comment = False)

File: msal/src/Microsoft.Identity.Client/AuthenticationResult.cs [hunk 12]
```diff
@@ -39,28 +39,33 @@ public class AuthenticationResult
     {
         private const string Oauth2AuthorizationHeader = "Bearer ";
         private readonly MsalAccessTokenCacheItem _msalAccessTokenCacheItem;
+        private readonly MsalIdTokenCacheItem _msalIdTokenCacheItem;
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/AuthenticationResult.cs]: >_msalIdTokenCacheIte [](start = 46, length = 20)

as far as I can see in most other places we do not use _ / can we pls. stick to using either or? (personal preferences is without the _)

**Comment by henrik-me** [comment]
>         internal readonly object LockObject = new object();

Pls. add a story as part of the overall cache work to review our current locking strategy. We have seen a number of requests on that and there are good suggestions on optimizations. Also it should be considered if we should have an even more granular persistence/serialization design.


---
Refers to: msal/src/Microsoft.Identity.Client/TokenCache.cs:60 in 0a4acbb. [](commit_id = 0a4acbb7026204a4698a1b012eb19ec2f45648e6, deletion_comment = False)

**Comment by henrik-me** [comment]
>     public static class TokenCacheExtensions

I do not see something like this in ADAL. Can that be added?

---
Refers to: msal/src/Microsoft.Identity.Client/Features/TokenCache/TokenCacheExtensions.cs:41 in 0a4acbb. [](commit_id = 0a4acbb7026204a4698a1b012eb19ec2f45648e6, deletion_comment = False)

File: msal/tests/dev apps/DesktopTestApp/MsalUserRefreshTokenControl.cs [hunk 13]
```diff
@@ -33,14 +35,16 @@ public MsalUserRefreshTokenControl()
 
         private void InvalidateRefreshTokenBtn_Click(object sender, System.EventArgs e)
         {
-            _item.RefreshToken = GarbageRtValue;
-            _cache.SaveRefreshTokenCacheItem(_item);
+            _rtItem.Secret = GarbageRtValue;
+            _cache.SaveRefreshTokenCacheItem(_rtItem, null);
```
Comment (henrik-me) [inline msal/tests/dev apps/DesktopTestApp/MsalUserRefreshTokenControl.cs]: >ull); [](start = 55, length = 5)

you have lots of these... consider having an override that doesn't require this to be specified instead.

File: msal/tests/Test.MSAL.NET.Unit/UnifiedCacheTests.cs [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        public void UnifedCache_MsalStoresToAndReadRtFromAdalCache()
+        {
+            PublicClientApplication app = new PublicClientApplication(TestConstants.ClientId);
+
+            app.UserTokenCache.legacyCachePersistance = new TestLegacyCachePersistance();
+
+            MockWebUI ui = new MockWebUI()
+            {
+                MockResult = new AuthorizationResult(AuthorizationStatus.Success,
+                    TestConstants.AuthorityHomeTenant + "?code=some-code")
```
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/UnifiedCacheTests.cs]: >?code=some-code" [](start = 57, length = 16)

TestConstant ? (same below)

File: msal/tests/Test.MSAL.NET.Unit/UnifiedCacheTests.cs [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+                {
+                    {"grant_type", "refresh_token"}
+                },
+                ResponseMessage =
+                    MockHelpers.CreateSuccessTokenResponseMessage(TestConstants.UniqueId,
+                        TestConstants.DisplayableId,
+                        TestConstants.Scope.ToArray())
+            });
+
+            // Using RT from Adal cache for silent call
```
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/UnifiedCacheTests.cs]: >RT [](start = 21, length = 2)

Are you sure it's the RT and not the AT?

File: msal/tests/Test.MSAL.NET.Unit/UnifiedCacheTests.cs [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+            AuthenticationResult result1 = app.AcquireTokenSilentAsync
+                (TestConstants.Scope, result.User, TestConstants.AuthorityCommonTenant, false).Result;
+
+            Assert.IsNotNull(result1);
```
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/UnifiedCacheTests.cs]: don't you need to validate that the result returned is in more details what is expected? (btw. same comment above)

File: msal/tests/Test.MSAL.NET.Unit/RequestsTests/InteractiveRequestTests.cs [hunk 17]
```diff
@@ -149,18 +150,17 @@ public void NoCacheLookup()
...
                 TokenType = "Bearer",
                 ExpiresOnUnixTimestamp = CoreHelpers.DateTimeToUnixTimestamp(DateTime.UtcNow + TimeSpan.FromSeconds(3599)),
                 ScopeSet = TestConstants.Scope
             };
-            atItem.IdToken = IdToken.Parse(atItem.RawIdToken);
-            atItem.ClientInfo = ClientInfo.CreateFromJson(atItem.RawClientInfo);
-            MsalAccessTokenCacheKey atKey = atItem.GetAccessTokenItemKey();
-            atItem.AccessToken = atKey.ToString();
-            cache.TokenCacheAccessor.AccessTokenCacheDictionary[atKey.ToString()] = JsonHelper.SerializeToJson(atItem);
+            atItem.CreateDerivedProperties();
```
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/RequestsTests/InteractiveRequestTests.cs]: >CreateDerivedProperties(); [](start = 19, length = 26)

as mentioned a few other places. not a fan of this

**Review by henrik-me — CHANGES_REQUESTED** [review]
:clock1:

File: adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs (hunk 1 above)
Comment (SomkaPe) [reply adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs]: we want to request it for all flows like in Msal, in AddAdditionalRequestParameters we add flow specific parameters

File: adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs (hunk 2 above)
Comment (SomkaPe) [reply adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs]: to not interfere with AuthorityType from Core

File: msal/src/Microsoft.Identity.Client/IUser.cs (hunk 3 above)
Comment (SomkaPe) [reply msal/src/Microsoft.Identity.Client/IUser.cs]: will do after confirming exact list of claims to be returned as User -> Account

File: msal/src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs (hunk 5 above)
Comment (SomkaPe) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs]: According to new schema AT, RT and IdTokens are called credentials and all of them have Secret - which is basically content of token

Comment (SomkaPe) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs]: so just to be more consistent with new schema

File: msal/tests/dev apps/XForms/XForms/RefreshTokenCacheItemDetails.xaml.cs (hunk 6 above)
Comment (SomkaPe) [reply msal/tests/dev apps/XForms/XForms/RefreshTokenCacheItemDetails.xaml.cs]: will change

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs (hunk 7 above)
Comment (SomkaPe) [reply msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs]: it is possible, but since many properties (including key properties) are derived from idToken and client Info , CacheItem is almost useless without derived props

File: core/src/Cache/MsalCacheItemBase.cs (hunk 10 above)
Comment (SomkaPe) [reply core/src/Cache/MsalCacheItemBase.cs]: it is internal , not a part of api

File: core/src/Cache/MsalIdTokenCacheItem.cs (hunk 11 above)
Comment (SomkaPe) [reply core/src/Cache/MsalIdTokenCacheItem.cs]: in many test cases cache items are created manually so , it has to be called , so it is internal

File: core/src/Cache/MsalRefreshTokenCacheItem.cs [hunk 18]
```diff
@@ -25,68 +25,57 @@
...
     {
-
-        public MsalRefreshTokenCacheItem()
+        internal MsalRefreshTokenCacheItem()
         {
+            CredentialType = Cache.CredentialType.RefreshToken.ToString();
         }
-
-        public MsalRefreshTokenCacheItem(string environment, string clientId, MsalTokenResponse response) : base(clientId)
+        internal MsalRefreshTokenCacheItem(string environment, string tenantId, string clientId, MsalTokenResponse response) : this()
```
[the remark this answers is not in the document]
Comment (SomkaPe) [reply core/src/Cache/MsalRefreshTokenCacheItem.cs]: sorry did not get the question, please explain more

File: core/src/Cache/MsalRefreshTokenCacheItem.cs [hunk 19]
```diff
@@ -25,68 +25,57 @@
...
-                ClientInfo = ClientInfo.CreateFromJson(RawClientInfo);
-            }
+            return new MsalAccountCacheKey(Environment, TenantId, UserIdentifier).ToString();
         }
+
+        [DataMember(Name = "target")]
+        internal string Scopes { get; set; }
+
+        [DataMember(Name = "realm")]
+        internal string TenantId { get; set; }
```
[the remark this answers is not in the document]
Comment (SomkaPe) [reply core/src/Cache/MsalRefreshTokenCacheItem.cs]: will do

File: adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs (hunk 1 above)
Comment (henrik-me) [reply adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs]: the name of the method doesn't indicate that to me. Also is this really the only special parameter? As far as I can see you will get the same if you move this line to the Additional method. And for other places including the additionalparams they will also include this request param.

---
In reply to: [191035079](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1065#discussion_r191035079) [](ancestors = 191035079)

File: adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs (hunk 2 above)
Comment (henrik-me) [reply adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs]: Can't you use the namespace to be more specific instead? Also why do we not only have this in core? do we need to have this type in other places?

---
In reply to: [191035342](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1065#discussion_r191035342) [](ancestors = 191035342)

File: msal/src/Microsoft.Identity.Client/IUser.cs (hunk 3 above)
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/IUser.cs]: ok

---
In reply to: [191035389](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1065#discussion_r191035389) [](ancestors = 191035389)

File: msal/src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs (hunk 5 above)
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs]: thanks

---
In reply to: [191035446](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1065#discussion_r191035446) [](ancestors = 191035446)

File: msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs (hunk 7 above)
Comment (henrik-me) [reply msal/tests/Test.MSAL.NET.Unit/PublicClientApplicationTests.cs]: imo: this complexity should be hidden inside the class. This code is too fragile imo

---
In reply to: [191035568](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1065#discussion_r191035568) [](ancestors = 191035568)

File: core/src/Cache/MsalCacheItemBase.cs (hunk 10 above)
Comment (henrik-me) [reply core/src/Cache/MsalCacheItemBase.cs]: single responsibility. external code from the class is calling this, which is not good. If we really need this for what ever optimization, then it should be wrapped into a helper class that does this upon initialization or similar.

---
In reply to: [191035584](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1065#discussion_r191035584) [](ancestors = 191035584)

File: core/src/Cache/MsalIdTokenCacheItem.cs (hunk 11 above)
Comment (henrik-me) [reply core/src/Cache/MsalIdTokenCacheItem.cs]: it is not internal. internal to the class. as mentioned other places the complexity can be hidden in a helper class but right now it's all over the place.

---
In reply to: [191035635](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1065#discussion_r191035635) [](ancestors = 191035635)

File: core/src/Cache/MsalRefreshTokenCacheItem.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (henrik-me) [reply core/src/Cache/MsalRefreshTokenCacheItem.cs]: I believe I saw additional constructors earlier.. must have been dreaming. I wanted you to call other contructors using this instead of just the base constructor.

---
In reply to: [191035701](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1065#discussion_r191035701) [](ancestors = 191035701)

File: adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs (hunk 1 above)
Comment (SomkaPe) [reply adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs:321]: it is not only special parameter , requestParameters is initialized with resourcem, clientId

Comment (SomkaPe) [reply adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs:321]: so this params are common for all flows, and AddAditionalRequestParameters is used to set flow specific params

File: adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs (hunk 2 above)
Comment (SomkaPe) [reply adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs:446]: do you want me to add Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Instance ?
Enums have different values so it is not easy to join them in one
